c10/cuda/CUDACachingAllocator.cpp

```diff
@@ -454,10 +461,77 @@ struct ExpandableSegment {
     return rangeFromHandles(begin, end);
   }
 
+  // Setup IPC sharing for range.
+  // Returns the (larger) range that was actually shared.
+  // Serializes data to std::ostream that can be passed to the
+  // other process, and then restored as an exapandable segment
+  // via ExpandableSegment::fromShared(istream);
+  SegmentRange share(SegmentRange range, std::ostream& buf) {
+    auto begin = segmentLeft(range.ptr);
+    auto end = segmentRight(range.ptr + range.size);
+    ShareHeader header{getpid(), segment_size_, end - begin};
+    buf.write((const char*)&header, sizeof(ShareHeader));
+    for (auto i : c10::irange(begin, end)) {
+      auto& handle = handles_.at(i).value();
+      if (!handle.fd) {
+        int fd = 0;
+        C10_CUDA_DRIVER_CHECK(DriverAPI::get()->cuMemExportToShareableHandle_(
```

Comment:
Couple of questions for my understanding (pardon my noob understanding)

- Does obtaining exportable handle increment the refcount of the backing handle? (i.e. does it need to be closed explicitly?).
- Are we overwriting the previous fd (original one obtained with cuMemCreate) below with newly obtained fd?

Reply:
The export creates a file descriptor. I do not think the cuMemCreate command actually creates a file descriptor. It only tells the library that one may be requested in the future.